File: scripts/types/ingredients.ts
import { Unit, EditableWithId, Ingredient, IngredientPrice } from "./common";

export interface IngredientWithPrice extends Ingredient {
    price: IngredientPrice;
}

export interface IngredientsViewModel {
    product_names: { [productId: string]: string };
    ingredients: IngredientWithPrice[];
    units: Unit[];
}

export interface IngredientExtended extends IngredientWithPrice, EditableWithId {
    isBase: boolean;
    displayPrice: string;
    displayQuantity: string;
    unit: Unit;
}

export interface IngredientsData extends IngredientsViewModel {
    init: () => void;

    newIngredientType: string;

    ingredients_ext: IngredientExtended[];
    ingredientBackup: Record<number, IngredientExtended>;

    setIngredientPrice(ingredient: IngredientExtended): void
    setIngredientQuantity(ingredient: IngredientExtended): void
    getFilteredUnitsForUnitId(unitId: number): Unit[]

    startEditing: (usage: IngredientExtended) => void;
    cancelEditing: (usage: IngredientExtended) => void;
    removeItem: (itemId: number) => void;
    modifyIngredient(ingredient: IngredientWithPrice): IngredientExtended;
}

File: viewModels/ingredients.go
package viewmodels

import "github.com/mike-jl/price_calc/db"

type IngredientWithPrices struct {
	Ingredient db.Ingredient        `json:"ingredient"`
	Prices     []db.IngredientPrice `json:"prices"`
}

type IngredientWithPrice struct {
	ID    int64              `json:"id"`
	Name  string             `json:"name"`
	Price db.IngredientPrice `json:"price"`
}

type IngredientsViewModel struct {
	Ingredients  []IngredientWithPrice `json:"ingredients"`
	Units        map[int64]db.Unit     `json:"units"`
	ProductNames map[int64]string      `json:"product_names"`
}
